MarkLogic support for JSON documents

wemrysi opened the pull request:
Adds support for querying and uploading data as JSON in MarkLogic.

A new query parameter, `format`, is recognized in MarkLogic connection URIs, indicating which kind of documents to operate on. This is intended to work in concert with #1766 to provide a way to mount segments of XML or JSON data in the database.

This is a large refactoring of the MarkLogic planner that, roughly, involves
1. Refactoring effects to use typeclass constraints in APIs and treating `Free` as an implementation detail
2. Abstracting out the structural operations for MarkLogic into `StructuralPlanner`, leveraging this elsewhere in the planner and adding a phantom type index to the MarkLogic `Planner` typeclass to allow indexing over the document format.
3. Implementing `StructuralPlanner` for both XML and JSON and update `Data` encoding/decoding to support both formats.
4. Adding a format phantom type index to the MarkLogic `WriteFile` interpreter used to index over `AsContent` instances.

Closes #1704.

wemrysi commented on effect/src/main/scala/quasar/effect/Failure.scala:
```diff
@@ -62,6 +63,14 @@ object Failure {
       def apply[A](v: EitherT[F, E, A]): F[A] = unattempt(v.run)
     }
 
+    implicit val monadError: MonadError[F, E] =
```
It's defined in `LiftedOps`.

sellout replied:
Ah, yes … this is one of those cases we have to fix, where something that _looks_ like a type parameter isn’t. We should’t name any aliases things like `M`, `F`, `A` …. I know there is a bad precedent somewhere in the Mongo connector, but it’s very confusing.

wemrysi replied:
Yeah, agreed. I'm the guilty party that introduced the offending alias. I'll add a `FIXME` to the file as a reminder as changing it will affect many dependencies.

wemrysi commented on common/src/main/scala/quasar/common/package.scala:
```diff
@@ -17,11 +17,13 @@
   type PhaseResults = Vector[PhaseResult]
   type PhaseResultW[A] = Writer[PhaseResults, A]
   type PhaseResultT[F[_], A] = WriterT[F, PhaseResults, A]
+  type PhaseResultTell[F[_]] = MonadTell_[F, PhaseResults]
```
In theory, yes, but there are still plenty of places that reference `PhaseResultT`, only a couple that use `PhaseResultW`.

wemrysi commented on marklogic/src/main/scala/quasar/physical/marklogic/fs/MarkLogicConfig.scala:
```diff
@@ -0,0 +1,76 @@
+
+  object DocType {
+    case object Json extends DocType
+    case object Xml  extends DocType
```
Good point, `DocType` was one of the last things added to the PR and I hadn't thought of unifying it with `fmt`.

wemrysi commented on marklogic/src/main/scala/quasar/physical/marklogic/fs/data.scala:
```diff
@@ -86,6 +112,66 @@ object data {
+      case EJsType(DT.NA) => (Data.NA: Data).point[F]
+
+      // NB: Because sometimes we need to encode `Null` as an object due to
+      //     MarkLogic losing map entries where the value is a literal `null`.
```
I still think the empty sequence is a better representation for `NA` than `null` as many (most?) operations on the empty seq result in the empty seq, it has cardinality 0, it isn't a `node()` (when type matching) whereas none of those things are true for `null-node{}`.

When casted, the encoded `null` will end up as a `null-node()` and is handled appropriately, the problem is during map concat as any keys with `null` values are lost during the merge. `null` values are otherwise preserved outside of merging.

sellout replied:
“any keys with null values are lost during the merge” is exactly the behavior we want for `Data.NA`. I mean, hopefully it’s already impossible for for `Data.NA` to get to a connector in a query, and if `Data.NA` ever comes out in a result set (which it shouldn’t), it’s now removed before we send the response to the user.

But that also means that it probably doesn’t matter what representation you use for `Data.NA`, and thankfully there will be no such thing once we’re using EJson.

wemrysi replied:
I think I was talking past you a bit as I keep conflating `Undefined` and `Data.NA`, apologies. I thought the former was the `QScript` representation of the latter, is that not the case?

sellout replied:
Yeah, so they’re not _too_ different. Except that `Undefined` has no value representation. `Data.NA` was originally created to allow for partial failures when reading data. E.g., if we get back a BSON Regex value, we have no `Data` for that, so it’s `Data.NA`[†]. But there’s also an “operational” notion of partial failure (`3 + "s"`), and this is `Undefined`, which we’ve largely been using Data.NA for. However, the semantics of Undefined are such that you never end up with such a value. I.e.
* `MakeMap("key", Undefined) ≟ {}`,
* `MakeMap(Undefined, "value") ≟ {}`,
* `MakeArray(Undefined) ≟ []`,
* `AddMetadata("value", Undefined) ≟ value`
* `IfUndefined(Undefined, "rescue") ≟ "rescue"`,
* `Cond(true, "pass", Undefined) ≟ "pass"`,
* `Cond(false, Undefined, "fail") ≟ "fail"`, and
* `Undefined` passed to any other operation (I _think_ I got all the cases) results in `Undefined`.

[†]: With EJson, we should have _some_ encoding of a BSON Regex, likely something like `"someRegex" @ { "_ejson.type": "_bson:regex"}`, so reading any value from the collection should succeed.

wemrysi commented on marklogic/src/main/scala/quasar/physical/marklogic/fs/data.scala:
```diff
@@ -19,18 +19,44 @@ package quasar.physical.marklogic.fs
 import scalaz.{Node => _, _}, Scalaz._
 
 object data {
+  val encodeJson: Data => Json = {
```
I found this encoding easier to work with as it uses consistent key names for all the different types and I needed more precise control over the temporal formats to make them easy to cast into XDM types.

For example, `LocalTime.parse("10:14:00").toString` results in the string `"10:14"` (which is what the precise encoding would output) which isn't directly castable as an `xs:time` because the latter requires the seconds field to be present, even if it is zero. There are similar caveats for the other temporal types.

wemrysi commented on it/src/main/resources/tests/matchRegexPatternWithExpressions.test:
```diff
@@ -16,7 +16,7 @@
   "predicate": "containsExactly",
 
-  "ignoreFieldOrder": ["couchbase"],
+  "ignoreFieldOrder": ["couchbase", "marklogic_json"],
```
The current implementation should preserve document order for a `select * from ...`, but that seems to be about it as the mechanism used to build maps doesn't seem to preserve order in all cases and I haven't figured out what it uses to determine the order.

However, there is another method of constructing objects that we might be able to use that does appear to preserve order. I originally started with this, but it isn't useful in general as it appears to be purely syntactic and so we can't use it when we don't statically know the shape of the map. Going this route would likely increase the scenarios where order is preserved, but there will still exist those where it isn't (i.e. adding a new key -> value association or merging with another map).

I'll give it a quick go, but may just end up filing an issue for this, in the interest of time for the release, if it looks like it is a more involved change.

sellout replied:
Yeah, I’m fine with postponing this. Honestly, I’m not too concerned about it until we have the fallback evaluator. At that point, it basically means leaving a `Map` node for fbeval to process over the result stream, but we want to only do that when it’s necessary, so as much as possible, we want to be able to tell when the query isn’t going to be able to preserve field order.

wemrysi commented:
@sellout I've addressed most of your comments either with more comments or commits. Let me know what you think about the `Data` encodings and I'll try and deal with the field ordering first thing tomorrow (though, if you're amenable to it I'd almost prefer just adding an issue and addressing it in a separate PR).

sellout commented:
Annoyingly, there is now a merge conflict.

wemrysi commented:
@sellout Annoying indeed, merged.
